server/authz/authz.go

```diff
@@ -70,16 +70,14 @@ func (a *Authorizer) SkipAuthorization(ctx context.Context) {
 	}
 }
 
-// IsAlreadyAuthorized returns true if the context has already authorized. This
-// may be useful to skip authorization checks in the Service layer when e.g.
-// the current request has already been authorized due to its different
-// authentication method, such as the device authentication token which allows
-// reading the corresponding host's information even if there is no user
-// associated with the request (so typical user-based authorization checks
-// would fail).
-func (a *Authorizer) IsAlreadyAuthorized(ctx context.Context) bool {
+// IsAuthenticatedWith returns true if the request has been authenticated with
+// the specified authentication method, false otherwise. This is useful to avoid
+// calling Authorize if the request is authenticated with a method that doesn't
+// support granular authorizations - provided it is ok to grant access to the
+// protected data.
+func (a *Authorizer) IsAuthenticatedWith(ctx context.Context, method authz_ctx.AuthenticationMethod) bool {
```

Comment:
The old `IsAlreadyAuthorized` call was not good enough, especially with the addition of a call to `svc.AppConfig` that needs to work from a device-authenticated request. `IsAlreadyAuthorized` considered any `SkipAuth` or `Authorize` call as being already authorized, but a single request may call `Authorize` to check for many things - e.g. the permission to read hosts, and then to read app_config. Just because it has already been authorized does not mean more checks are not required.

So `IsAuthenticatedWith` fixes this by checking instead the authentication method of the request, and it should be called explicitly to skip `Authorize` calls only where needed, so that it is a deliberate thing in the code and is only done at places where we know we have to allow authorization when a request is authenticated with the specified method.